sgkit/io/vcf/vcf_writer.py

```diff
@@ -49,220 +46,227 @@ def dataset_to_vcf(
-
     with ExitStack() as stack:
         if isinstance(output, str) or isinstance(output, Path):
             output = stack.enter_context(open(output, mode="w"))
 
+        header_str = input.attrs["vcf_header"]
+
         print(header_str, end="", file=output)
 
-        if n_variants == 0:
+        if input.dims["variants"] == 0:
             return
 
         header_info_fields = _info_fields(header_str)
         header_format_fields = _format_fields(header_str)
 
-        # fixed fields
-
-        contigs = np.array(ds.attrs["contigs"], dtype="S")
-        filters = np.array(ds.attrs["filters"], dtype="S")
-
-        chrom = ds.variant_contig.values
-        pos = ds.variant_position.values
-        id = ds.variant_id.values.astype("S")
-        alleles = ds.variant_allele.values.astype("S")
-        qual = ds.variant_quality.values
-        filter_ = ds.variant_filter.values
-
-        # info fields
+        contigs = np.array(input.attrs["contigs"], dtype="S")
+        filters = np.array(input.attrs["filters"], dtype="S")
 
-        # preconvert all info fields to byte representations
-        info_bufs = []
-        info_mask = np.full((len(header_info_fields), n_variants), False, dtype=bool)
-        info_indexes = np.zeros(
-            (len(header_info_fields), n_variants + 1), dtype=np.int32
-        )
-
-        k = 0
-        info_prefixes = []  # field names followed by '=' (except for flag/bool types)
-        for key in header_info_fields:
-            var = f"variant_{key}"
-            if var not in ds:
-                continue
-            if ds[var].dtype == np.bool:
-                values = ds[var].values
-                info_mask[k] = create_mask(values)
-                info_bufs.append(np.zeros(0, dtype=np.uint8))
-                # info_indexes contains zeros so nothing is written for flag/bool
-                info_prefixes.append(key)
-                k += 1
-            else:
-                values = ds[var].values
-                if values.dtype.kind == "O":
-                    values = values.astype("S")  # convert to fixed-length strings
-                info_mask[k] = create_mask(values)
-                info_bufs.append(
-                    np.empty(vcf_values_to_byte_buf_size(values), dtype=np.uint8)
-                )
-                vcf_values_to_byte_buf(info_bufs[k], 0, values, info_indexes[k])
-                info_prefixes.append(key + "=")
-                k += 1
-
-        info_mask = info_mask[:k]
-        info_indexes = info_indexes[:k]
-
-        info_prefixes = np.array(info_prefixes, dtype="S")
+        for ds in _variant_chunks(input):
+            dataset_chunk_to_vcf(
+                ds, header_info_fields, header_format_fields, contigs, filters, output
+            )
 
-        # format fields
 
-        # these can have different sizes for different fields, so store in sequences
-        format_values = []
-        format_bufs = []
+def dataset_chunk_to_vcf(
```

Comment:
This looks like a bigger change than it actually is, as it's basically just an indentation change.